lib/backend/etcdv3/watcher.go

```diff
@@ -0,0 +1,203 @@
+
+	// We use wc.ctx to reap all goroutines. Under whatever condition, we should stop them all.
+	// It's fine to double cancel.
+	wc.cancel()
+	close(wc.resultChan)
+}
+
+// Stop() implements the api.WatchInterface.
+// This calls through to the context cancel function.
+func (wc *watcher) Stop() {
+	wc.cancel()
+}
+
+// ResultChan() implements the api.WatchInterface.
+func (wc *watcher) ResultChan() <-chan api.WatchEvent {
+	return wc.resultChan
+}
+
+// listCurrent retrieves the existing entries and sends an event for each listed
+func (wc *watcher) listCurrent() error {
+	log.Info("Performing initial list with no revision")
+	list, err := wc.client.List(wc.list, "")
+	if err != nil {
+		return err
+	}
+
+	wc.initialRev, err = strconv.ParseInt(list.Revision, 10, 64)
+	if err != nil {
+		log.WithError(err).Error("List returned revision that could not be parsed")
+		return err
+	}
+
+	for _, kv := range list.KVPairs {
+		log.Info("Sending create events for each existing entry")
+		wc.sendEvent(&api.WatchEvent{
+			Type: api.WatchAdded,
+			New:  kv,
+		})
+	}
+	return nil
+}
+
+// watchLoop starts a watch on the required path prefix and sends a stream of
+// event updates for internal processing.
+func (wc *watcher) watchLoop(watchClosedCh chan struct{}) {
+	log.Debug("Starting watcher.watchLoop")
+	if wc.initialRev == 0 {
+		// No initial revision supplied, so perform a list of current configuration
+		// which will also get the current revision we will start our watch from.
+		if err := wc.listCurrent(); err != nil {
+			log.Errorf("failed to list current with latest state: %v", err)
+			wc.sendError(err)
+			return
+		}
+	}
+	opts := []clientv3.OpOption{clientv3.WithRev(wc.initialRev + 1), clientv3.WithPrevKV()}
+
+	// If we are not watching a specific resource then this is a prefix watch.
+	key := model.ListOptionsToDefaultPathRoot(wc.list)
+	logCxt := log.WithFields(log.Fields{
+		"etcdv3-key": key,
+		"rev":        wc.initialRev,
+	})
+	logCxt.Debug("Starting etcdv3 watch")
+	if !model.ListOptionsIsFullyQualified(wc.list) {
+		logCxt.Debug("Performing prefix watch")
+		opts = append(opts, clientv3.WithPrefix())
+		key += "/"
+	}
+	wch := wc.client.etcdClient.Watch(wc.ctx, key, opts...)
+	for wres := range wch {
+		if wres.Err() != nil {
+			err := wres.Err()
+			log.WithError(err).Error("Watch channel error")
+			wc.sendError(err)
+			return
+		}
+		for _, e := range wres.Events {
+			if ae, err := convertWatchEvent(e, wc.list); ae != nil {
+				wc.sendEvent(ae)
+			} else if err != nil {
+				wc.sendError(err)
+			}
+		}
+	}
+
+	// The watch has been ended through a client action (e.g context cancelled or timedout).
+	//
+	// When we come to this point, it's only possible that client side ends the watch.
+	// e.g. cancel the context, close the client.etcdClient.
+	close(watchClosedCh)
+}
+
+// sendError sends an error down the errChan or waits for done notification.
+func (wc *watcher) sendError(err error) {
+	select {
```

Comment:
again, we prolly don't need `select{}` since both channel operations below are blocking

Reply:
so what would this look like if you didn't use a select?